### PR TITLE
[EXPERIMENT] Reduction of monomorphization load through type erasure

### PR DESCRIPTION
**Pull request by cjgillot**
This is an early proposition in response to #77960 through type-erasure of a function's MIR. This is still an experiment, and there are probably a lot of edge cases I did not cover. Any weird addition to the test cases is welcomed :smile:

Here is the description I gave in https://github.com/rust-lang/rust/issues/77960#issuecomment-821857686. Consider a function

```rust
fn foo<T>(a: M<T>) -> R<T> {
  let c: C<T> = T::CONST;
  // stuff with T
  T::bar(a, c)
}
```

During MIR analyses, the trampoline and the callee version can be computed generically:
* gather all constants inside MIR body, and replace them by appropriately-typed parameters;
* gather all functions, and replace them by function pointer parameters.

```rust
fn foo<T>(a: M<T>) -> R<T> {
  let _2: C<T> = T::CONST;
  let _3: fn(M<T>, C<T>) -> R<T> = T::bar;
  foo_impl<T>(a, _2, _3)
}

fn foo_impl<T>(a: M<T>, c: C<T>, _3: fn(M<T>, C<T>) -> R<T>) -> R<T> {
  // stuff with T
  _3(a, c)
}
```

The next step is to only have one instance of foo_impl. In rustc parlance, I introduced an new variant `InstanceDef::ErasedShim` to encode it. It is resolved when some a MIR call terminator has the new flag `erased = true`. In order to define the substitutions for this instance, we replace all the type parameters by `u8`. The code becomes:

```rust
type T_impl = u8; // We do not need it to satisfy any trait except Sized.

fn foo<T>(a: M<T>) -> R<T> {
  let _2: C<T> = T::CONST;
  let _3: fn(M<T>, C<T>) -> R<T> = T::bar;
  transmute(foo_impl(transmute(a), transmute(_2), transmute(_3)))
}

fn foo_impl(a: M<T_impl>, c: C<T_impl>, _3: fn(M<T_impl>, C<T_impl>) -> R<T_impl>) -> R<T_impl> {
  // stuff with T_impl
  _3(a, c)
}
```

For this transformation to be sound, we need to check that all those transmutes are safe.

The problem is that Rust layouts all monomorphized types independently: in general `M<T>` and `M<T_impl>` have no reason to have a similar layout if T != U. In special cases, this can happen, for instance if `M<T>` only points to sized T through a reference.

As a consequence, we need to check that `M<T>` and `M<T_impl>` and all recursively accessible types have the same layout. Likewise for `C<T>` and `R<T>`. If any such test fails, the type-erasure is unsound, we need to abort and go back to full monomorphization. Actually, we should be able to restrict the check to all types that are involved in the MIR: types of locals, of accessed fields, of dereferenced pointers...

The implementation actually does this in reverse. First, we check the layout in the function `gather_types`, and bail out if there is an incompatibility. Second, we gather and replace all the constants in the MIR. Handling drops requires to convert them into calls to `drop_in_place` in the same step. Third, we build the caller MIR. Finally, we erased all the type parameters in the callee MIR.

cc @jumbatm @jyn514

### CODE REVIEW

File: compiler/rustc_mir/src/interpret/terminator.rs [hunk 1]
```diff
@@ -425,6 +433,9 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 // recurse with concrete function
                 self.eval_fn_call(drop_fn, caller_abi, &args, ret, unwind)
             }
+            ty::InstanceDef::ErasedShim(_) => {
+                bug!("Interpreting type-erased functions is not implemented.")
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/terminator.rs:437]: If this actually lands, I'd appreciate help making this code work in Miri. :)
Also, if the CTFE engine does not support these calls, we must make sure they are rejected in `const fn`.

Comment (cjgillot) [reply compiler/rustc_mir/src/interpret/terminator.rs:437]: If this ever lands, these calls will only be created by a MIR-opt pass. As a consequence, they should not appear in a `const fn`.

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/terminator.rs:437]: Even optimized code must, IMO, be executable by the Miri engine -- and indeed it is possible to run Miri on optimized MIR with `-Zmir-opt-level=4`. The code here is relevant well beyond `const fn`. So while it is okay to temporarily land a feature without Miri engine support, this is always a bug that should be fixed eventually.

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/terminator.rs:437]: But I guess this means we indeed don't have to reject anything in `const fn`. However, we only recently stopped optimizing the MIR used by CTFE, and we might revert that decision at some point, so there's a danger of some lock-in here.

**Comment by jackh726** [comment]
I think this definitely needs an MCP and someone with more knowledge in this area to review.

**Comment by Mark-Simulacrum** [comment]
Going to go ahead and close as an inactive "experiment", but of course feel free to reopen if you make further progress or this becomes ready for review -- I agree with @jackh726 that a MCP may be warranted.

**Pull request closed without merging**